api/incubator/src/main/java/io/opentelemetry/api/incubator/common/ArrayBackedExtendedAttributesBuilder.java

```diff
@@ -36,11 +47,126 @@ public <T> ExtendedAttributesBuilder put(ExtendedAttributeKey<T> key, T value) {
     if (key == null || key.getKey().isEmpty() || value == null) {
       return this;
     }
+    if (key.getType() == ExtendedAttributeType.VALUE && value instanceof Value) {
+      putValue(key, (Value<?>) value);
+      return this;
+    }
     data.add(key);
     data.add(value);
     return this;
   }
 
+  private void putValue(ExtendedAttributeKey<?> key, Value<?> valueObj) {
+    // Convert VALUE type to narrower type when possible
+    String keyName = key.getKey();
+    switch (valueObj.getType()) {
+      case STRING:
+        data.add(stringKey(keyName));
+        data.add(valueObj.getValue());
+        return;
+      case LONG:
+        data.add(longKey(keyName));
+        data.add(valueObj.getValue());
+        return;
+      case DOUBLE:
+        data.add(doubleKey(keyName));
+        data.add(valueObj.getValue());
+        return;
+      case BOOLEAN:
+        data.add(booleanKey(keyName));
+        data.add(valueObj.getValue());
+        return;
+      case ARRAY:
+        @SuppressWarnings("unchecked")
+        List<Value<?>> arrayValues = (List<Value<?>>) valueObj.getValue();
+        ExtendedAttributeType attributeType = attributeType(arrayValues);
+        switch (attributeType) {
+          case STRING_ARRAY:
+            List<String> strings = new ArrayList<>();
+            for (Value<?> v : arrayValues) {
```

Comment:
If we're interested in premature performance optimization, there's probably a way to to merge this with `attributeType` function to iterate only once.

I'd say only pursue if it ends up making the code more concise with the perf improvement.